### DISCUSSION TITLE
Refactor script type system

### DISCUSSION THREAD

**Discussion by dr-orlovsky** [rust-bitcoin/rust-bitcoin#921] [Refactoring]
The need to refactor script type system was originally created by the fact that SegWit v0 prohibited use of uncompressed public keys, which led to the introduction of script contexts and complex error handling logic downstream in `rust-miniscript`. Now, with the introduction of TapScript and `XOnlyPublicKey`s the issue has become even more severe: #920.

Related discussion: https://github.com/rust-bitcoin/rust-bitcoin/issues/653

To start the discussion, here is my initial concept proposal for the "perfect" type system (not sure can we refactor into it here though):
```rust
/// Marker trait for script instruction sets
pub trait InstructionSet {
    fn op_code(&self) -> u8;
    fn takes_data(&self) -> usize;
    fn data(&self) -> Option<&[u8]>;
}

pub enum LegacyInstr {/* ... */}
pub enum Bip342Instr {/* ... */}

// # Script trait and iterators

pub struct InstrIter<'script, S> {
    script: &'script S,
    offset: usize,
}

pub trait Script<Instr>:
    IntoIterator<Item = Instr> + FromIterator<Instr> + FromStr + Display
{
    fn instructions(&self) -> InstrIter<Self>;
    fn bytes(&self) -> core::slice::Iter<u8>;
}

// # Script base types defining specific sets of instructions

pub struct LegacyScript(Box<u8>);
impl Script<LegacyInstr> for LegacyScript {
    /*...*/
}

pub struct SegwitV0Script(Box<u8>);
impl Script<LegacyInstr> for SegwitV0Script {
    /*...*/
}

pub trait LeafScript {
    fn version(&self) -> LeafVersion;
}

pub struct TapScript(Box<u8>);
impl Script<Bip342Instr> for TapScript {
    /*...*/
}
impl LeafScript for TapScript {
    /*...*/
}

pub struct FutureLeafScript {
    script: Box<u8>,
    version: FutureLeafVersion,
}
// we do not implement Script here since we can't parse future script versions
impl LeafScript for FutureLeafScript {
    /*...*/
}

/// Not actually a script but a value stored in `TxOut::script_pubkey`
pub enum ScriptPubkey {
    Bare(LegacyScript),
    Pkh(PubkeyHash),
    Sh(ScriptHash),
    Wpkh(WPubkeyHash),
    Wsh(WScriptHash),
    Tr(OutputPubkey),
    Future {
        version: WitnessVersion,
        payload: Box<u8>,
    },
}

/// Type which is used to generate `ScriptPubkey` of transaction output.
pub enum ScriptPubkeyDescr {
    Bare(LegacyScript),
    Pkh(LegacyPubkey), // compressed or uncompressed
    Sh { redeem_script: LegacyScript },
    Wpkh(CompressedPubkey),
    Wsh { witness_script: SegwitV0Script },
    ShWpkh(CompressedPubkey),
    ShWsh { witness_script: SegwitV0Script },
    Tr { internal_key: InternalPubkey, merkle_root: Option<TapTreeNode> },
}

pub enum TapTreeNode {
    Leaf(Box<dyn LeafScript>),
    Hidden(sha256::Hash),
    Branch {
        left: Box<TapTreeNode>,
        right: Box<TapTreeNode>,
    }
}
```

Pubkey types here are taken from https://github.com/rust-bitcoin/rust-bitcoin/discussions/919

**Comment by apoelstra** [comment]
What would the type in `TxOut` and `TxIn` be?

**Reply by dr-orlovsky** [reply]
```rust

pub struct TxOut { value: Satoshis, script_pubkey: ScriptPubkey }

pub struct TxIn { prevout: OutPoint, sig_script: SigScript, witness: Witness, seq_no: u32 }

pub struct SigScript(LegacyScript)
```

**Comment by apoelstra** [comment]
Ah, good observation that `script_sig` is always a legacy script. Actually, so is the `script_pubkey` -- actual segwit/tapleaf scripts appear only in witness data, which we already understand as `Vec<Vec<u8>>` when it is stored inside witness data. (There is a plausible argument for having a richer `Witness` type which has a script and other stack elements ... but I think that is likely to be more complication than it's worth.)

So we shouldn't have a dedicaded `ScriptPubkey` type (and actually, nack that type regardless of its usefulness...that sort of templating logic belongs in rust-miniscript).

**Reply by dr-orlovsky** [reply]
So what `script_pubkey` type should be? From my understanding (a) it is not a script but rather set of instructions with very specific interpretation by consensus rules (for instance regarding the witness version, which interpretation is clearly not part of general scripting) and (b) clearly should be more structured than just `Vec<u8>`. Thus, I see no alternative to templating, which is not about miniscript and internal script structure at all, but only about "script pubkey descriptor" structuring which can be used separatly from miniscript and more like address payload structure (except it is bit more generic covering bare scripts).

**Reply by dr-orlovsky** [reply]
Regarding the witness type we may have something like

```rust
pub struct TxIn {
    prevout: OutPoint,
    seq_no: u32,
    satisfaction: Satisfaction
}

pub enum Satisfaction {
    Legacy(ScriptSig),
    SegwitV0(Witness),
    TaprootKey(schnorr::Signature),
    TaprootScript(TaprootWitness),
    Future { version: WitnessVersion, witness: Witness }
}

pub struct TaprootWitness {
    control_block: ControlBlock,
    script: LeafScript,
    annex: Vec<u8>,
    stack: Witness
}

```

**Reply by apoelstra** [reply]
`script_pubkey` is a script, and it is interpreted according to the legacy rules. Anything else would represent a hardfork of Bitcoin.

Yes, there are additional verification steps that were introduced by P2SH, segwit, etc. We don't care about those. This library does not do script interpretation.

**Reply by dr-orlovsky** [reply]
@apoelstra, AFAIK `script_pubkey` is interpreted with different op-code execution basing on the height of the block the transaction is coming from: pre-segwit, `script_pubkeys` starting with `0` byte will push an empty element into the stack (and with v1 `0x51` will push 0x01 into the stack); while post-segwit this push is not happening. So this is _differently interpreted_ code, meaning different scripting language.

Since we do not have an access to the block height value inside our current library, we can't interpret whether `script_pubkey` is pre-segwit `LegacyScript` or post-segwit "segwit scriptPubkey" and we need a type marking that uncertainty: `pub struct ScriptPubkey(Script)` at least. This newtype should not check anything on its creation (it can take any byte sequence to construct) and will not interpret the meaning of the script; but unlike simple `LegacyScript`/`Script` must provide API methods like `is_p2tr`, `new_p2tr` and others. Callers to this methods must ensure that they are using them in the proper context.

**Comment by sanket1729** [comment]
Script is just a collection of bytes. As @apoelstra mentioned even scripts like `<XonlyPublicKey> Checksig` are valid scripts in segwit context. They will verify, but it is not rust-bitcoin to decide. For example, we could a such a script in an unexecuted branch of `OP_IF`. Or even a compressed key in segwit context if unexecuted is perfectly fine! Templating is not only hard to do but also incorrect. 

`rust-miniscript` can get away with assuming these because it is specifically a subset of scripts that make sense. It will not able to parse things from blockchain that are incomplete pushes, or any weird things. But it will be a bug in `rust-bitcoin` if it cannot parse a transaction from the bitcoin blockchian. 

What we can do a dedicated type system so that users don't use witness script to compute a legacy address etc. But I am not sure if it is worth the effort. 

```rust
struct ScriptPubkey(Script); //
struct SegwitV0WitnessScript(Script); // constructor would check the size limit 10,000 bytes
struct ScriptSig(Script); // 
LegacyScript(Script); // p2sh() would then accept Legacy Script instead of a general script
```

Then we can have APIs for standardness sanity checks
1) Check that ScriptSig consists only of pushes
2) Check that the script is complete and iterates correctly (no incomplete pushes)
3) Specific checks regarding disabled opcodes/ op_success opcodes 

I am not sure if it is worth it to all this, or if it even belongs to rust-bitcoin. We can have `script_utils` or `script_extras` to these things. I personally don't see too much utility in this because we already have `rust-miniscript`. 

Maybe, it is a part of me that does not want to review/redo the things that are already done cleanly in rust-miniscript.

**Reply by dr-orlovsky** [reply]
Thank you for detailed rationale.

I think there are at least two (or three) points of view, defining criteria for designing a bitcoin library.

The first one is consensus support and parsing existing blockchain data. Networking, validation, block explorers. And what you and Andrew say mostly goes into this basket.

Second one is a point of view of a wallet, interested in composing a standard transaction - an finding/parsing standard transactions which belong to it.

Third one is a non-wallet transaction modifiers (like those related to client-side-validation protocols - opentimestamps, RGB, other future single-use-seal applications - to which data a transaction may commit to). These are still within the standardness rules, but significantly diverge from simple wallet transaction composers (for instance, they may hold bitcoin-related data structures offchain and use different validation rules for them; or modify taproot script trees in an unusual way like with OP_RETURN taproot commitments).

Our difference arises from the fact that it seems this library is primarily focused on the first use case, while my interest lies is in the third one. And the second one is sort of ground in between I am also interested in, but no one is really addressing it in this library (yet). 

Another thing is that I can't rely on miniscript in full, since miniscript does not work for existing LN transactions. It also can't be used for client-side-validation, since it is still not finalized and can be modified/updated in the future.

**Reply by dr-orlovsky** [reply]
Ah, there is a forth point of view: LN. There are different criteria, like Matt mentioning the need to get rid of keeping secp pubkeys inside node ids, otherwise parsing network topology becomes really under-performant. These are examples of non-transaction use of bitcoin primitives outside of the consensus or even standardness scope.

**Reply by sanket1729** [reply]
> Another thing is that I can't rely on miniscript in full, since miniscript does not work for existing LN transactions. It also can't be used for client-side-validation, since it is still not finalized and can be modified/updated in the future.

I don't know much about client-side validation. But you can use rust-miniscript with an `HTLC` descriptor that you can implement downstream with `DesriptorTrait`. I do not see any reason why this should not work. Seeing that lightning is so common, we can even merge it in `rust-miniscript`, what do you say @apoelstra?

**Reply by dr-orlovsky** [reply]
@sanket1729 this will not add ability to build scripts from miniscript, but just a descriptor support, which is a different topic (but also nice to have).

**Reply by sanket1729** [reply]
> this will not add ability to build scripts from miniscript

What functionality do you want here? Do you mean parsing a descriptor from `bitcoin::Script`?

**Reply by dr-orlovsky** [reply]
@sanket1729, not sure what you call "descriptor" here... I always thought that the descriptor is the external part of the story, describing how specific script (which may be written/templated in different languagues) is encoded into `scriptPubkey` and how its satisfaction is encoded into `sigScript` and `witness`. The internal script of the descriptor can be bitcoin script hex bytes, OP_CODES, miniscript, something else (in the future).

So by saying "LN is non-compatible to miniscript" I am not touching descriptors at all: with them I have no problem here. It is just `wsh(OP_IF .... OP_CSV ...)`. What I mean by incompatibility is the inability to put miniscript inside such descriptor. And adding "HTLC" descriptor makes little sense, since it is not "descriptor" at all: if you write `htlc(...)` or `wsh(htlc(...))` this will be just a dirty hack killing all abstractions between descriptor and script (or you may say that you invent a new scripting language dedicated for a single-case LN, next to bitcoin op-codes and miniscript - but why do we need it if we can just use opcodes/hex encoding?)

**Reply by dr-orlovsky** [reply]
> The internal script of the descriptor can be bitcoin script hex bytes, OP_CODES, miniscript, something else (in the future).

that is why I do not understand why descriptors reside in `miniscript` and not `miniscript` is based on generalzied descriptors. And that is why I am using my own descriptor library: https://crates.io/crates/descriptors

(the case of LN goes here: https://docs.rs/descriptors/0.5.0/descriptors/struct.ScriptTemplate.html)

**Reply by dr-orlovsky** [reply]
> I don't know much about client-side validation. 

This is when you put the commitment to some extra-transaction data into transaction (with P2C, S2C, taproot-OP_RETURN etc) and keep all the state & scripts on the client-side. The owner of some state execute the script and validates it - and no body else in the world needs to know those data or re-do that job. Blockchain is used just as a consensus layer, not data/script layer.

Huge storage efficiency. Much less validation overhead. Great privacy. Arbitrary complex scripting. Amazing scalability (you can put client-side-validation over LN: you host commitment in transaction, which may be an off-chain tx inside state channel).

Invented by Peter Todd (opentimestamps is the first simple case of it; more had followed - single-use-seals, RGB). Designed and developed at LNP/BP Association (https://github.com/LNP-BP) by Giacomo Zucco, me and a lot of other chaps.

What are the applications? Private & scalable smart contracts over LN (RGB). BiFi, up to algorithmic stable coins, DEX, decentralized trustless storage and computing network (paid with bitcoin over LN) - not requiring even a single soft-fork / new op code / any changes to Bitcoin Core

**Reply by sanket1729** [reply]
> What I mean by incompatibility is the inability to put miniscript inside such descriptor

Do you mean random policies composed in an htlc? This is not allowed by LN-spec
As an end-user, you should not care about miniscript, only about descriptors. As long as all the descriptor APIs work for you, you need to even know if there is a miniscript being used internally. 

> if you write htlc(...) or wsh(htlc(...)) this will be just a dirty hack killing all abstractions between descriptor and script (

I think this is useful to abstractly reason about things and have systematic backups. I think this is useful because users can use descriptors that will do all the ugly translation to script under the hood. It is useful for new wallets that want to implement LN and don't want to implement all the raw byte code with script templates. One of the points of rust-miniscript/descriptors is to "Do all the ugly things once" so that we never have to do them again.

**Reply by sanket1729** [reply]
> The internal script of the descriptor can be bitcoin script hex bytes, OP_CODES, miniscript, something else (in the future).

If the descriptor spec if updated to support any of these, rust-miniscript will be updated to support these.

**Reply by dr-orlovsky** [reply]
> Do you mean random policies composed in an htlc?

No. I mean representing lightning commitment transaction outputs as descriptors with miniscript inside.

> As an end-user, you should not care about miniscript, only about descriptors. 

I am not an "end-user" :)

> It is useful for new wallets that want to implement LN and don't want to implement all the raw byte code with script templates. One of the points of rust-miniscript/descriptors is to "Do all the ugly things once" so that we never have to do them again.

But this will mean changing miniscript language (breaking its guarantees) or breaking compatibility with Bitcoin Core descriptors / standard. Or are you talking about "yet another language inside of descriptors"? I doubt LN node devs will appreciate it, since you will have a need to constantly update it after BOLT changes (happening constantly), i.e. become part of the LN libs.

**Reply by dr-orlovsky** [reply]
> If the descriptor spec if updated to support any of these, rust-miniscript will be updated to support these.

descriptor spec does not support miniscript yet :)

**Comment by apoelstra** [comment]
> But it will be a bug in rust-bitcoin if it cannot parse a transaction from the bitcoin blockchian.

This is true, but we can still parse scripts *differently* (e.g. labeling the opcodes differently) in different contexts. They just have to round-trip as bytestrings.

I'm fine with adding script contexts to rust-bitcoin. As I observed above, this wouldn't complicate much of anything, since the main transaction types would just use the legacy Script type. The non-legacy ones wouldn't really be connected to anything, though we could have e.g. a `script_pubkey` method from a Segwit script which would output a legacy script of the form `00 <scripthash>` which would be cool.

Regarding the different points of view, @dr-orlovsky I think you are correct that we care about consensus here and "base layer" stuff (and several of us also care about wallet development, but we care over at rust-miniscript, not here) and generally tend to ask extra functionality be moved into other crates. Though this isn't such a hard/fast rule ... you could argue that e.g. bip32 support doesn't belong here, for example, but it's so pervasive throughout the ecosystem.

**Reply by dr-orlovsky** [reply]
That's why I think that PSBT and BIP32 deserves their own crates: together with rust-minisdcript they form a "wallet" layer. Also I have a lot of wallet-specific extensions in descriptor-wallet lib (made on top of miniscript and the rest) which can't fit in rust-bitcoin - but can be easily migrated to dedicated crates like bip32.
